src/Microsoft.AspNetCore.Dispatcher/DispatcherValueEndpointSelector.cs

```diff
@@ -2,54 +2,244 @@
+                    //
+                    // `entries` will appear in `OrdinalIgnoreCaseEntries` exactly once and in `OrdinalEntries` once
+                    // for each variation of casing that we've seen.
+                    if (!OrdinalEntries.ContainsKey(routeValues))
+                    {
+                        OrdinalEntries.Add(routeValues, entries);
+                    }
+                }
+            }
+
+            public int Version { get; }
+
+            public string[] RouteKeys { get; }
+
+            public Dictionary<string[], List<SimpleEndpoint>> OrdinalEntries { get; }
+
+            public Dictionary<string[], List<SimpleEndpoint>> OrdinalIgnoreCaseEntries { get; }
+        }
+
+        private class StringArrayComparer : IEqualityComparer<string[]>
+        {
+            public static readonly StringArrayComparer Ordinal = new StringArrayComparer(StringComparer.Ordinal);
+
+            public static readonly StringArrayComparer OrdinalIgnoreCase = new StringArrayComparer(StringComparer.OrdinalIgnoreCase);
+
+            private readonly StringComparer _valueComparer;
+
+            private StringArrayComparer(StringComparer valueComparer)
+            {
+                _valueComparer = valueComparer;
+            }
+
+            public bool Equals(string[] x, string[] y)
+            {
+                if (object.ReferenceEquals(x, y))
+                {
+                    return true;
+                }
+
+                if (x == null ^ y == null)
+                {
+                    return false;
+                }
+
+                if (x.Length != y.Length)
+                {
+                    return false;
+                }
+
+                for (var i = 0; i < x.Length; i++)
+                {
+                    if (string.IsNullOrEmpty(x[i]) && string.IsNullOrEmpty(y[i]))
+                    {
+                        continue;
+                    }
+
+                    if (!_valueComparer.Equals(x[i], y[i]))
                     {
                         return false;
                     }
                 }
+
+                return true;
+            }
+
+            public int GetHashCode(string[] obj)
+            {
+                if (obj == null)
+                {
+                    return 0;
+                }
+
+                var hash = new HashCodeCombiner();
+                for (var i = 0; i < obj.Length; i++)
+                {
+                    hash.Add(obj[i], _valueComparer);
+                }
+
+                return hash.CombinedHash;
             }
+        }
 
-            return true;
+        private void EnsureServices(HttpContext context)
+        {
+            var loggerFactory = (ILoggerFactory)context.RequestServices.GetService(typeof(ILoggerFactory));
+            _logger = loggerFactory.CreateLogger<DispatcherValueEndpointSelector>();
```

Comment:
Use the `GetRequiredService<T>` method.

You can get `ILogger<DispatcherValueEndpointSelector>` directly